include/deal.II/grid/grid_tools.h

```diff
@@ -556,6 +556,100 @@ namespace GridTools
+   *
+   * @note This function may return both locally owned and ghost cells, but
+   * is expected not to return artificial cells.
+   *
+   * @tparam Container A type that satisfies the requirements of a mesh
+   * container (see @ref GlossMeshAsAContainer).
+   * @param[in] container A mesh container (i.e. objects of type Triangulation,
+   * DoFHandler, or hp::DoFHandler).
+   * @param[in] predicate A function defining the subdomain around which the
+   * halo layer is to be extracted. It is a function that takes in an active cell
+   * and returns a boolean.
+   * @return A list of active cells sharing at least one common vertex with the
+   * predicated subdomain.
+   *
+   * @author Jean-Paul Pelteret, Denis Davydov, Wolfgang Bangerth, 2015
+   */
+  template <class Container>
+  std::vector<typename Container::active_cell_iterator>
+  compute_active_cell_halo_layer (const Container                                                                    &container,
+                                  const std_cxx11::function<bool (const typename Container::active_cell_iterator &)> &predicate);
+
+  /**
+   * Extract and return ghost cells which are the active cell layer
+   * around all locally owned cells. This is most relevant for
+   * parallel::shared::Triangulation where it may return a subset of all
+   * ghost cells on a processor, but for parallel::distributed::Triangulation
+   * this should return all the ghost cells.
+   *
+   * @tparam Container A type that satisfies the requirements of a mesh
+   * container (see @ref GlossMeshAsAContainer).
+   * @param[in] container A mesh container (i.e. objects of type Triangulation,
+   * DoFHandler, or hp::DoFHandler).
+   * @return A list of ghost cells
+   *
+   * @author Jean-Paul Pelteret, Denis Davydov, Wolfgang Bangerth, 2015
+   */
+  template <class Container>
+  std::vector<typename Container::active_cell_iterator>
+  compute_ghost_cell_halo_layer (const Container &container);
+
+#ifdef DEAL_II_WITH_CXX11
+
+  /**
+   * Extract and return all active cells around a subdomain defined by a
+   * collection of material ids.
+   *
+   * @tparam Container A type that satisfies the requirements of a mesh
+   * container (see @ref GlossMeshAsAContainer).
+   * @param[in] container A mesh container (i.e. objects of type Triangulation,
+   * DoFHandler, or hp::DoFHandler).
+   * @param[in] materials A collection of material ids that define the
+   * subdomain that is encapsulated by the cell layer that is to be extracted.
+   * @return The list of cells that comprise the halo layer
+   *
+   * @note This function may return both locally owned and ghost cells, but
+   * is expected not to return artificial cells.
+   *
+   * @author Jean-Paul Pelteret, Denis Davydov, Wolfgang Bangerth, 2015
+   */
+  template <class Container>
+  std::vector<typename Container::active_cell_iterator>
+  compute_active_cell_halo_layer (const Container                    &container,
```

Comment:
Out of curiosity, do you have a specific need for the halo defined by material ids?